src/puppetlabs/puppetdb/query_eng/engine.clj

```diff
@@ -263,7 +263,9 @@
                                        [:= :certnames.certname :fs.certname]
 
                                        :reports
-                                       [:= :certnames.latest_report_id :reports.id]
+                                       [:and
+                                        [:= :certnames.certname :reports.certname]
+                                        [:= :certnames.latest_report_id :reports.id]]
```

Comment:
Since this is just a backport of the actual fix, it should probably be done separately. Will file a ticket.